thinc/backends/ops.py

```diff
@@ -680,6 +682,167 @@ def backprop_relu(
         dY *= Y > 0
         return dY
 
+    def relu_n(
+        self, X: FloatsType, n: float = 6.0, inplace: bool = False
+    ) -> FloatsType:
+        if inplace:
+            self.xp.clip(X, 0, n, out=X)
+            return X
+        return self.xp.clip(X, 0, n)
+
+    def backprop_relu_n(
+        self, dY: FloatsType, X: FloatsType, n: float = 6.0, inplace: bool = False
+    ) -> FloatsType:
+        if inplace:
+            dY *= (0 < X) & (X < n)
+            return dY
+        return dY * ((0 < X) & (X < n))
+
+    # Following https://www.scitepress.org/Papers/2019/74696/74696.pdf
+    def hard_sigmoid(self, X: FloatsType, inplace: bool = False) -> FloatsType:
+        if inplace:
+            X *= 0.2  # type: ignore
+            X += 0.5  # type: ignore
+            return cast(FloatsType, self.xp.clip(X, 0, 1, out=X))
+        out = X * 0.2 + 0.5
+        return cast(FloatsType, self.xp.clip(out, 0, 1))
+
+    def backprop_hard_sigmoid(
+        self, dY: FloatsType, X: FloatsType, inplace: bool = False
+    ) -> FloatsType:
+        slope = self.alloc1f(1) + 0.2
```

Comment:
I am not sure if I understand why this needs an array? Also, since `alloc1f` creates a `float32` array by default, this would reduce precision if `X` is a `float64` array.

Reply:
When the result of the activation goes propagates to `thinc.backends.numpy_ops.NumpyOps.gemm` on the backward pass I've gotten this error, when returning `float64`: 

```
File "blis/py.pyx", line 64, in blis.py.gemm
ValueError: Buffer dtype mismatch, expected 'const double' but got 'float'
```
So my understanding was that `thinc` ops should live in `float32`

Reply:
Do you have a reproducer for this error. I think that this might happen when the input is float32, but when you outputting float64, because then `gemm` will be called with matrices of different types. I think it should work correctly if when the inputs are float32, you return float32 and when the inputs are float64, you return float64.

The issue is that you loose the type with the conversion to bools and then the result gets whatever the result of `slope` is. If you do something like

```
dX = self.xp.where(self.xp.abs(X) < 2.5, 0.2 * X, 0.)
```

`dX` will always have the same dtype as `X`.

Reply:
With the new `clipped_linear` do you think it would make sense to replace the `relu`, `relu_n` and `hard_sigmoid` with just a `ClippedLinear` layer that calls `clipped_linear` in the `ops`?

Reply:
I pushed a `ClippedLinear` layer, if you change the `Relu` in `examples/mnist.py` to `ClippedLinear` it will throw the 

```
  File "thinc/backends/numpy_ops.pyx", line 93, in thinc.backends.numpy_ops.NumpyOps.gemm
    return blis.py.gemm(x, y, out=out, trans1=trans1, trans2=trans2, beta=0.)
  File "blis/py.pyx", line 64, in blis.py.gemm
ValueError: Buffer dtype mismatch, expected 'const double' but got 'float'
```

Reply:
See my comment above, I think this happens if `gemm` is passed arrays of different types. Things like

```
dX = slope * ((low < X) & (X < high))
```

Are not safe, since they discard the data type of X and result in an array that has the same data type as `slope`. E.g., if `slope` is a Python `float`, this fill create a `np.float64` array. However, this is incorrect when the input as a `np.float32` array. Then in the caller of the backward function, a `gemm` is attempted of an `np.float32` and an `np.float64` array.

See my earlier comment for an approach (with `np.where`) that should not have this issue.

Reply:
By the way, the current implementation should also not typecheck, since `dY`, `X` and the return type must be the same type (`FloatsType` is a generic type).

Reply:
Oh okay sorry! Because of you typecheck comment @danieldk I've remembered that both `numpy` and `cupy`has `zeros_like` so I've implemented `clipped_linear` with that, because I kinda like that pattern. The `hard_sigmoid` and `relu_n` are now calling `clipped_linear` and its `backprop` and we have `hard_tanh` now too.